socketio/src/asynchronous/socket.rs

```diff
@@ -0,0 +1,228 @@
+use super::generator::StreamGenerator;
+use crate::{
+    error::Result,
+    packet::{Packet, PacketId},
+    Error, Event, Payload,
+};
+use async_stream::try_stream;
+use bytes::Bytes;
+use futures_util::{Stream, StreamExt};
+use rust_engineio::{
+    asynchronous::Client as EngineClient, Packet as EnginePacket, PacketId as EnginePacketId,
+};
+use std::{
+    fmt::Debug,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+#[derive(Clone)]
+pub(crate) struct Socket {
+    //TODO: 0.4.0 refactor this
+    engine_client: Arc<EngineClient>,
+    connected: Arc<AtomicBool>,
+    generator: StreamGenerator<Packet>,
+}
+
+impl Socket {
+    /// Creates an instance of `Socket`.
+    pub(super) fn new(engine_client: EngineClient) -> Result<Self> {
+        let connected = Arc::new(AtomicBool::default());
+        Ok(Socket {
+            engine_client: Arc::new(engine_client.clone()),
+            connected: connected.clone(),
+            generator: StreamGenerator::new(Self::stream(engine_client, connected)),
```

Comment:
nit: Cloning this doesn't seem safe, might result in undefined behavior. We could end up with two independent clients. Might cause issues for secure socket io websocket for instance

Reply:
Cloning an object only guarantees a copy of the present state, but depending on the implementation of clone/and the struct the behavior could either be:
- Two rust objects pointing to the same data using arc/locks
- Two rust objects that are copies of the same data when they're cloned, but changes to one isn't reflected into the other (if we don't have guarantees that engine_client does this it could result in strange behavior)

Reply:
That's true, for the engine_client we know that this works as all necessary parts live behind an `Arc`. But I see the smell, this is not obvious. I will think about another way.

Reply:
Good idea, for now I added a test and updated documentation. But I think I'll change that in the future.